external/asf4-drivers/qtouch/qtouch.c

```diff
@@ -517,50 +517,43 @@ static uint16_t sensor_previous_filtered_reading[DEF_NUM_SENSORS][DEF_SENSOR_NUM
 /* Custom sensor signal filter. */
 uint16_t qtouch_get_sensor_node_signal_filtered(uint16_t sensor_node)
 {
-    // Filter the sensor signal using the equation:
-    //     Y = X / (1 + 0.0001 * X^2)
+    // Filter the sensor signal.
     //
-    // The maximum Y value is 50 when X = 100, with a decreasing tail for X > 100.
-    // The equation can be plotted using: https://www.desmos.com/calculator
-    // and is an approximation of `Y = arctan(X)` but scaled in Y.
+    // Smooth it out and saturate it so that values never go beyond 50.
+    // This helps to mitigate 'jumpy' channels that exist at higher sensor readings when
+    // in noisy environments.
     //
-    // Smaller sensor readings will have higher relative weights, improving sensitivity of
-    // neighboring channels for better linearity and less deadband, while helping to mitigate
-    // 'jumpy' channels that exist at higher sensor readings when in noisy environments.
-    //
-    // The above empirically worked better than a sigmoid function Y = X / (1 + 0.015 * X).
-    //
-    uint8_t j;
-    double X, Y;
-    uint16_t Y_ave;
-    X = qtouch_get_sensor_node_signal(sensor_node) - qtouch_get_sensor_node_reference(sensor_node);
-    X = (X < 0) ? 0 : X;
+    uint16_t X;
+    uint16_t sensor_raw = qtouch_get_sensor_node_signal(sensor_node);
+    uint16_t sensor_reference = qtouch_get_sensor_node_reference(sensor_node);
+    X = sensor_raw < sensor_reference ? 0 : sensor_raw - sensor_reference;
     // Add more weight to edge buttons because they are physically smaller (smaller readings).
     if ((sensor_node == DEF_SCROLLER_OFFSET_0) ||
         (sensor_node == DEF_SCROLLER_OFFSET_1) ||
         (sensor_node == DEF_SCROLLER_OFFSET_0 + DEF_SCROLLER_NUM_CHANNELS - 1) ||
         (sensor_node == DEF_SCROLLER_OFFSET_1 + DEF_SCROLLER_NUM_CHANNELS - 1)) {
         X = X * (1 + DEF_SENSOR_EDGE_WEIGHT);
     }
-    Y = X / (1 + 0.0001 * X * X);
+    // Saturate out-of-range readings.
+    X = X > 50 ? 50 : X;
 
     // Calculate sensor readout using a moving average
     // The moving average wieghts previous N readings twice current reading
     uint16_t moving_average_cummulative_weight = 1;// Add one for current reading calculated above
-    Y_ave = (uint16_t) Y;
-    for (j = 0; j < DEF_SENSOR_NUM_PREV_POS; j++) {
+    uint16_t X_ave = X;
+    for (int j = 0; j < DEF_SENSOR_NUM_PREV_POS; j++) {
```

Comment:
`j` should be an `usize` since it is used in indexing.

Reply:
As in `size_t`?

I'd rather have `int`s as much as possible in this simple situations, they make the code simpler to read and less of a mental overhead (imho). Happy to swap to `size_t` or similar if we like it better.

Reply:
Fixed.